cirq/contrib/acquaintance/gates.py

```diff
@@ -0,0 +1,311 @@
+                    acquaintance_gate=pre_acquaintance_gate,
+                    qubits=left_part[::-1],
+                    before=True,
+                    layers=layers,
+                    mapping=mapping)
+
+        layers['intra'].append(shift)
+        shift.gate.update_mapping(mapping, qubits)
+
+        # after
+        if ((left_size == acquaintance_size - 1) and
+            (right_size > 1)):
+            # right part
+            post_acquaintance_gate = ACQUAINT(*qubits[-acquaintance_size:])
+
+            new_left_part = qubits[right_size - 1::-1]
+            acquaint_insides(
+                    swap_gate=swap_gate,
+                    acquaintance_gate=post_acquaintance_gate,
+                    qubits=new_left_part,
+                    before=False,
+                    layers=layers,
+                    mapping=mapping)
+
+        if ((right_size == acquaintance_size - 1) and
+            (left_size > 1)):
+            # left part
+
+            post_acquaintance_gate = ACQUAINT(*qubits[:acquaintance_size])
+            acquaint_insides(
+                    swap_gate=swap_gate,
+                    acquaintance_gate=post_acquaintance_gate,
+                    qubits=qubits[right_size:],
+                    before=False,
+                    layers=layers,
+                    mapping=mapping)
+
+
+class SwapNetworkGate(CompositeGate, PermutationGate):
+    """A single gate representing a generalized swap network.
+
+    Args:
+        part_lens: An sequence indicating the sizes of the parts in the
+            partition defining the swap network.
+        acquaintance_size: An int indicating the locality of the logical gates
+            desired; used to keep track of this while nesting. If 0, no
+            acquaintance gates are inserted.
+
+    Attributes:
+        part_lens: See above
+        acquaintance_size: See above.
+        swap_gate: The gate used to swap logical indices.
+    """
+
+    def __init__(self,
+                 part_lens: Sequence[int],
+                 acquaintance_size: int=0,
+                 swap_gate: Gate=SWAP
+                 ) -> None:
+        if len(part_lens) < 2:
+            raise ValueError('part_lens must have length at least 2.')
+        self.part_lens = tuple(part_lens)
+        self.acquaintance_size = acquaintance_size
+        self.swap_gate = swap_gate
+
+
+    def default_decompose(self, qubits: Sequence[QubitId]) -> OP_TREE:
+        qubit_to_position = {q: i for i, q in enumerate(qubits)}
+        mapping = {q: i for i, q in enumerate(qubits)}
+        parts = []
+        q = 0
+        for part_len in self.part_lens:
+            parts.append(list(qubits[q: q + part_len]))
+            q += part_len
+        n_parts = len(parts)
+        op_sort_key = (lambda op:
+                qubit_to_position[min(op.qubits, key=qubit_to_position.get)] %
+                self.acquaintance_size)
+        layers = {'posterior_interstitial': []
```

Comment:
*definitely* use a `namedtuple`

You can use `typing.NamedTuple` to add on type information